unittest/src/fake/syscalls/yield_impl.rs

```diff
@@ -52,5 +50,30 @@ pub(super) fn yield_wait() {
         return;
     }
 
-    unimplemented!("TODO: Implement upcalls");
+    if invoke_next_upcall() == libtock_platform::YieldNoWaitReturn::NoUpcall {
+        // yield-wait was called but there is no upcall queued. In a real Tock
+        // system, this process would be put to sleep until an upcall was queued
+        // (e.g. by an interrupt). However, in this single-threaded test
+        // environment, there is no possibility a new upcall will be enqueued
+        // while we wait. Therefore we instead panic; tests and code under test
+        // should never call yield-wait with no upcall queued.
+        panic!("yield-wait called with no queued upcall");
+    }
+}
+
+// Pops the next upcall off the kernel data's upcall queue and invokes it, or
+// does nothing if the upcall queue was entry. The return value indicates
+// whether an upcall was run. Panics if no kernel data is present.
+fn invoke_next_upcall() -> libtock_platform::YieldNoWaitReturn {
```

Comment:
It seems to me that this should just return a bool indicating whether an upcall was invoked. It seems strange to return `YieldNoWaitReturn` when this method is also called in response to yield-wait

Reply:
I used `YieldNoWaitReturn` because it had the values I needed, but I agree the name is confusing. I changed to `bool` as you suggested.

I also replaced the `if X { panic!(); }` pattern with `assert!(X, "...")`.